lib/asset_cloud/callbacks.rb

```diff
@@ -7,29 +7,41 @@ module Callbacks
     module ClassMethods
       def callback_methods(*symbols)
         symbols.each do |method|
-          code = <<-"end_eval"
-           def self.before_#{method}(*callbacks, &block)
-             callbacks << block if block_given?
-             write_inheritable_array(:before_#{method}, callbacks)
-           end
-
-           def self.after_#{method}(*callbacks, &block)
-             callbacks << block if block_given?
-             write_inheritable_array(:after_#{method}, callbacks)
-           end
-
-           def #{method}_with_callbacks(*args)
-             if execute_callbacks(:before_#{method}, args)
-               result = #{method}_without_callbacks(*args)
-               execute_callbacks(:after_#{method}, args)
-             end
-             result
-           end
-
-           alias_method_chain :#{method}, 'callbacks'
-          end_eval
-
-          self.class_eval code, __FILE__, __LINE__
+          define_callbacks(method)
+        end
+      end
+
+      def define_callbacks(method)
+        before = :"before_#{method}"
+        after = :"after_#{method}"
+        extension_module.send(:define_method, method) do |*args, &block|
```

Comment:
So that if you call `define_callbacks` multiple times, we re-use the same module. It's a quite common pattern.